### PR TITLE
New unit for NetHTTP 

### PR DESCRIPTION
**Pull request by p-samuel**
- A new unit was created packing the NetHTTP routines. This unit will be used as an alternative to the already existing Indy unit and will be instanciated in the API factory unit. The developer will be able to choose which library he wants to utilize to make the ntfy HTTP comunication by a compiler directive called `NTFY_HTTP_INDY`. 

``` pascal
Notify.Api.Factory

function TNotifyApiFactory.Api: INotifyApi;
begin
  {$IFDEF NTFY_HTTP_INDY}
  Result := TNotifyApiIndy.New
  {$ELSE}
  Result := TNotifyApiNetHTTP.New;
  {$IFEND}
end;
```

Without this compiler directive, the library will utilize NetHTTP by default. The developer can alter accessing `Project > Options > Building > Delphi Compiler > Conditional Defines` and adding `NTFY_HTTP_INDY `on the conditional defines list.

- `Sample `directory was renamed to `samples`, so that the dependency manager ignores this folder when installing, avoiding the units inside this directory to be added on the  project's search path.

- A new example was included demonstrating how utilizing this library on windows services.

- Two warnings were also removed at Notify.Api.Indy unit. They were caused by implicit cast operation when a string variable was assigned to a UTF8String variable.

``` pascal
LStrings := SplitString(UTF8ToString(LEventString) , #$A);

  for LString in LStrings do
    NxHorizon.Instance.Post<TNotifySubscriptionEvent>(Utf8String(LString));
```
- TNotifyResponseData was renamed to TNotifyResponseDTO.

- In case the developer opts to utilize Indy components, the openssl libraries will be loaded according to the operational sistem of the client. It has not yet been tested in some operational systems.

### CODE REVIEW

File: src/Notify.Api.Factory.pas [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+uses
+  Notify.Api.NetHTTP,
+  Notify.Api.Indy;
+
+{ TNotifyApiFactory }
+
+function TNotifyApiFactory.Api: INotifyApi;
+begin
+  {$IFDEF INDY}
```
Comment (rhatherall) [inline src/Notify.Api.Factory.pas:25]: This needs documenting in the README.md. It's effectively a part of your public interface "how to switch between Indy/NetHTTP".

Also, consider renaming this compiler define to `NTFY_INDY`. Consider, with such a generic define based on the name of a very popular network component set, what would happen if your library was included in other large projects that may have libraries with the same defines for other purposes? You may even consider having a more descriptive name `NTFY_DO_A_THING_WITH_INDY` so it becomes self-documenting when people include it in their projects.

Comment (p-samuel) [reply src/Notify.Api.Factory.pas:25]: Marking out for improvement. I'll make the necessary changes. I think `NTFY_INDY` is more descriptive name as you mentioned as it avoids clashing with other name that a large library might have. I'll also provide documentation of that in the README.md file.

Comment (p-samuel) [reply src/Notify.Api.Factory.pas]: @rhatherall, your review has been submitted. I will really appreciate if you have any other suggestions.

**Comment by p-samuel** [comment]
♻ Refactoring

- Updated conditional defines for switching between Indy and NetHTTP
- Updated readme documentation
- FreeAndNil removed from Api classes.

**Comment by p-samuel** [comment]
♻ NetHTTP Compatibility issues.

- There has been noticed some issues when installing this library on Delphi releases lower than 10.4 Sydney. 
- I'll proceed to refactor two points on the NetHTTP class that breaks when we try to install on these Delphi releases

``` pascal
function TNotifyApiNetHTTP.ClearHeaders: INotifyApi;
begin
  Result := Self;
  FNetHTTPRequest.CustHeaders.Clear; //does not exists in Delphi < 10.4 Sydney
end;
```

``` pascal
procedure TSSEThread.AbortStream;
begin
  FNetHTTPResquest.Cancel; //Does not exists in Delphi < 10.4 Sydney
end;
```

**Comment by p-samuel** [comment]
From product version 26 to 27, Embarcadero completed the NetHTTP class to perform an abort operation on current requests in transaction. Although many years has passed since the first release of NetHTTP, this class was incomplete and lacking these basic methods even up until that release- something that shocked me when I realized. This is a huge drawback if you are using Delphi 10.3 and lower. It seems they didn't have in mind that a client could possibly perform a long polling operation. 

Because of that, if we are using a product version lower or equal to 26 with NetHTTP, the library will be unable to unsubscribe from a topic unless we close and restart the application or service. Yes, huge drawback.

**Review by p-samuel — COMMENTED** [review]
Ready. I'm just working on a small example for Android subscriber.

**Comment by p-samuel** [comment]
Sorry guys, is being a long time but I finally got the opportunity to continue the project. This is not an example of how to follow a milestone. Anyhow, it seems that NetHTTP is now fully operational and maintaining long polling operations without issues.  As from my investigation the last year, I had several problems to perform this and even the "Cancel Request" method wasn't behaving as should. I'm not sure if Embarcadero did changes to NetHTTP this last version but seems very stable now. I was able to maintain more than 10h of long polling connection without timeout or hang. When I first saw that working was like melody to my ears and I could finally perform the tests I needed to Android. I prepared an example under ```samples``` folder containing the boilerplate to make foreground services run in Android. I didn't have any issues maintaining the long poll request and it worked very well in my tests.  The issue concerning reconection has also been addressed in this version. It was "a hack" but killed the bug with one shot.

Long story short, I'm happy to merge the PR with main and delete this branch once for all.
